sw/device/tests/entropy_src_fw_override_test.c

```diff
@@ -0,0 +1,407 @@
+    size_t len = words_to_input;
+    if (len > 32) {
+      len = 32;
+    }
+
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_nonblocking_read(
+        &entropy_src, input_buf, &len));
+
+    input_buf += len;
+    words_to_input -= len;
+  }
+
+  // If we've filled the input buffer and drained the output buffer, swap
+  // them.
+  if (words_to_input == 0 && words_to_output == 0) {
+    // Swap the buffers.
+    uint32_t *output_tmp = output_buf;
+    output_buf = input_buf - kFifoBufferSizeWords;
+    input_buf = output_tmp - kFifoBufferSizeWords;
+
+    words_to_input = kFifoBufferSizeWords;
+    words_to_output = kFifoBufferSizeWords;
+  }
+
+  if (words_to_output > 0) {
+    // Move out at most 16 samples into the override FIFO.
+    // This number has to be small otherwise the conditioner runs for too long
+    // and we spend too much time in this ISR.
+    size_t len = words_to_output;
+    if (len > 16) {
+      len = 16;
+    }
+
+    // Start the SHA3 block ready for us to start inserting into the FIFO.
+    CHECK_DIF_OK(dif_entropy_src_fw_override_sha3_start_insert(
+        &entropy_src, kDifToggleEnabled));
+
+    size_t words_written = 0;
+    CHECK_DIF_OK(dif_entropy_src_fw_ov_data_write(&entropy_src, output_buf, len,
+                                                  &words_written));
+
+    output_buf += words_written;
+    words_to_output -= words_written;
+
+    // Disable the SHA3 block, allowing it to finish processing and pass the
+    // data through to the output of the entropy source.
+    CHECK_DIF_OK(dif_entropy_src_fw_override_sha3_start_insert(
+        &entropy_src, kDifToggleDisabled));
+  }
+
+  CHECK_DIF_OK(dif_entropy_src_irq_acknowledge(
+      &entropy_src, kDifEntropySrcIrqEsObserveFifoReady));
+
+  CHECK_DIF_OK(dif_rv_plic_irq_complete(&rv_plic, kTopEarlgreyPlicTargetIbex0,
+                                        plic_irq_id));
+}
+
+static status_t configure_interrupts(void) {
+  TRY(dif_rv_plic_irq_set_priority(
+      &rv_plic, kTopEarlgreyPlicIrqIdEntropySrcEsObserveFifoReady, 0x1));
+
+  TRY(dif_rv_plic_target_set_threshold(&rv_plic, kTopEarlgreyPlicTargetIbex0,
+                                       0x0));
+
+  TRY(dif_rv_plic_irq_set_enabled(
+      &rv_plic, kTopEarlgreyPlicIrqIdEntropySrcEsObserveFifoReady,
+      kTopEarlgreyPlicTargetIbex0, kDifToggleEnabled));
+
+  TRY(dif_entropy_src_irq_set_enabled(
+      &entropy_src, kDifEntropySrcIrqEsObserveFifoReady, kDifToggleEnabled));
+
+  return OK_STATUS();
+}
+
+// Configure the entropy complex.
+static status_t entropy_config(
+    dif_entropy_src_single_bit_mode_t single_bit_mode,
+    bool bypass_conditioner) {
+  // Don't let the extract and insert interrupt handler run until we've enabled
+  // the entropy source.
+  irq_external_ctrl(false);
+
+  LOG_INFO("Disabling entropy complex");
+  // Disable the entropy complex.
+  TRY(entropy_testutils_stop_all());
+
+  // Disable all health tests.
```

Comment:
Maybe we should move this sequence to the test utils as well? It's useful and not complete obvious